src/nvs.rs

```diff
@@ -32,6 +32,91 @@ pub trait NvsPartitionId {
 
 pub struct NvsDefault(());
 
+/// A specialized key-value storage wrapper around `EspNvs` that provides a simplified interface
+/// for storing and retrieving arbitrary data as byte arrays.
+///
+/// This struct was introduced to solve issue [#585](https://github.com/esp-rs/esp-idf-svc/issues/585)
+/// where the `contains()` method in `EspNvs` incorrectly returned `false` for string values
+/// that actually existed in NVS partitions. The root cause was that `EspNvs` implements
+/// two different storage strategies:
+///
+/// 1. **Native ESP-IDF NVS API**: Direct access to ESP-IDF's native types (u8, u16, u32, u64, i8, i16, i32, i64, str, blob)
+/// 2. **Serialized storage**: Everything stored as either u64 (≤7 bytes) or blob (>7 bytes) for compatibility with serde
+///
+/// `EspKeyValueStorage` focuses on the second approach, providing a clean interface for:
+/// - Storing any data that can be represented as `&[u8]`
+/// - Automatic optimization: values ≤7 bytes stored as u64, larger values as blobs
+/// - Consistent `contains()` method that works correctly with this storage strategy
+/// - Full compatibility with Rust serde implementations (postcard, json, etc.)
+///
+/// ## Usage
+///
+/// ```rust,no_run
+/// use esp_idf_svc::nvs::{EspDefaultNvsPartition, EspKeyValueStorage};
+///
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let partition = EspDefaultNvsPartition::take()?;
+/// let storage = EspKeyValueStorage::new(partition, "my_namespace", true)?;
+///
+/// // Store data as bytes
+/// let data = b"hello world";
+/// storage.set_raw("my_key", data)?;
+///
+/// // Check if key exists (this works correctly, unlike the original EspNvs bug)
+/// assert!(storage.contains("my_key")?);
+///
+/// // Retrieve data
+/// let mut buffer = [0u8; 64];
+/// if let Some(retrieved) = storage.get_raw("my_key", &mut buffer)? {
+///     assert_eq!(retrieved, data);
+/// }
+/// # Ok(())
+/// # }
+/// ```
+///
+/// ## Performance Characteristics
+///
+/// - **Small values (≤7 bytes)**: Stored as u64 for efficiency
+/// - **Large values (>7 bytes)**: Stored as ESP-IDF blobs
+/// - **Memory efficient**: No unnecessary allocations for small values
+/// - **Flash efficient**: Optimized storage format reduces wear on flash memory
+pub struct EspKeyValueStorage<T: NvsPartitionId>(EspNvs<T>);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum NvsDataType {
+    U8 = 1,
```

Comment:
Perhaps instead of doing e.g. `U8 = 1` you could try `U8 = nvs_type_t_NVS_TYPE_U8 as_`?